README

Subject: Unifont archive version 1:5.1.20080706


OVERVIEW
--------
The Unifont is a dual-width (8x16/16x16) bitmap font, designed to provide
coverage for all of Unicode Plane 0, the Basic Multilingual Plane (BMP).
This version has a glyph for each visible code point in the Unicode 5.1
Basic Multilingual Plane.

Note that this version of the Unifont only provides a single glyph
for each character, making it impossible to handle any language that
needs context-dependent character shaping well. It is supplied in the
form of a hex file, with a converter to convert it to bdf. See
http://czyborra.com/unifont/ or http://unifoundry.com/unifont.html
for more information.

This is the unifoundry.com collection of utilities for the GNU
Unifont, assembled by Paul Hardy with the encouragement of the
font's creator, Roman Czyborra.  This archive contains the following
directories:

     dat       Reference versions of data files
     font      All the font information
     Makefile  The "make" file
     man       Unix man pages
     README    This file
     src       Source programs, in Perl and C

The original version of this README file was written by David Starner
and modified by Paul Hardy in 2008.


src/ AUTHORS
------------
Right now, all the Perl files in the src directory except "hex2sfd"
were written by Roman Czyborra (or in the case of johab2ucs2, by
Jungshik Shin, who then gave it to Roman).  Roman originally named
the "src/hexbraille" script as simply "braille"; Paul Hardy thought
there was too great a chance of a conflict with other utilities,
and so renamed it.  The "hex2bdf-split" Perl script in the source
directory is an adaptation of Roman's "hex2bdf" script that
David Starner wrote in 1999.

Luis Gonzalez Miranda wrote the original "hex2sfd" Perl script,
as well as a "howto-build.sh" shell script that Paul Hardy converted
into "./font/ttfsrc/Makefile".

All the C programs were written by Paul Hardy.

All Makefiles in this distribution were originally written by
Paul Hardy, and groomed for Debian by Anthony Fok.


unifont AUTHORS
---------------
Roman Czyborra created the original GNU Unifont, including the
.hex format.  For greater detail, see the HISTORY section below.

David Starner aggregated many glyphs contributed by others and
built these into earlier unifont releases.

Qianqian Fang began his Wen Quan Yi font in 2004; most of the
almost 30,000 CJK ideographs in the latest unifont were taken
from Wen Quan Yi with permission of Qianqian Fang.  The glyphs
in "./font/hexsrc/wqy-cjk.hex" are for the most part Qianqian
Fang's unibit and Wen Quan Yi glyphs.

Paul Hardy replaced the 11,172 thick-stroke Hangul Syllables glyphs
with thin-stroke glyphs (a desire expressed by Roman Czyborra for
years), merged Qianqian Fang's unibit and Wen Quan Yi glyphs into
the GNU Unifont (with lots of help and enthusiasm from Qianqian Fang),
drew about 8,500 more glyphs, and replaced the existing Tibetan glyphs
with new ones contributed by Rich Felker.


LICENSE
-------
Roman Czyborra released his work (Perl scripts and font .hex files)
under the following terms:

     All of my works you find here are freeware. You may
     freely copy, use, quote, modify or redistribute them
     as long as you properly attribute my contribution and
     have given a quick thought about whether Roman might
     perhaps be interested to read what you did with his
     stuff. Horizontal rules don't apply.

David Starner released what in this package is named "hex2bdf-split"
under the same license as Roman Czyborra's work.

License for all of Paul Hardy's work (except "johab2ucs2" and
"blanks.hex", mentioned separately), Makefile and debian/ mods
by Anthony Fok, and modified software from Luis Gonzalez Miranda
(with permission granted to Paul Hardy):

     These are released under the terms of the GNU General Public
     License version 2, or (at your option) a later version.

License for Fonts:

     Any fonts using glyphs from the "wqy-cjk.hex" file (including
     the default TrueType font) are bound by the terms of the Wen
     Quan Yi font license.  Those fonts are released under the terms
     of the GNU General Public License (GPL) versionn 2, with the
     exception that embedding the font in a document does not by
     itself bind that document to the terms of the GNU GPL.

     Any fonts that do not use glyphs from the "wqy-cjk.hex" file
     fall under the above "License for all of Roman Czyborra's work".

     The fonts in "./font/precompiled" do use wqy-cjk.hex, and so
     are licensed under the GNU GPL version 2, with the exception
     that embedding the font in a document does not in itself bind
     that document to the terms of the GNU GPL.  The following
     paragraphs explaining the exception is taken from the Wen
     Quan Yi font distribution:

          ** GPL v2.0 license with font embedding exception:

          As a special exception, if you create a document which
          uses this font, and embed this font or unaltered portions
          of this font into the document, this font does not by
          itself cause the resulting document to be covered by
          the GNU General Public License. This exception does not
          however invalidate any other reasons why the document
          might be covered by the GNU General Public License.
          If you modify this font, you may extend this exception
          to your version of the font, but you are not obligated
          to do so. If you do not wish to do so, delete this
          exception statement from your version.

License for "blanks.hex":

     There is one exception to the above rules: Paul Hardy earlier
     released the "blanks.hex" file into the public domain.


HISTORY
-------
Roman Czyborra <roman@czybora.com> began the GNU Unifont in 1998
as a low quality font to provide a glyph for every Unicode character
in the Basic Multilingual Plane.  He realized that no one font at
the time had complete coverage of the Unicode BMP.  Czyborra.com still
has several cool tools for the unifont not included here.

Since Roman Czyborra was unable to maintain the Unifont for a while,
and many patches existed on gnu-unifont@groups.yahoo.com
(http://groups.yahoo.com/group/gnu-unifont), David Starner
<dstarner98@aasaa.ofe.org> decided to make a new release extending the
Unifont with many characters in 1999.  That was the foundation of earlier
GNU Unifont releases in Debian.

By 2004, work had stopped on the GNU Unifont.  Qianqian Fang wanted to
create a high-quality Chinese Unicode font in 2004.  He began by copying
the GNU Unifont.  He replaced its Latin glyphs with another X11 font.
He replaced the existing main CJK ideographs with a higher quality font
that the People's Republic of China had placed in the public domain.
The new CJK glyphs in the ranges U+3400..U+4DB5 and U+4E00..U+9FA5
were developed from the Chinese National Standard GB19966-2005.

Qianqian named this new font "unibit", and released it under the terms
of the GNU General Public License (GPL) version 2, with the exception that
embedding his font in a document did not by itself bind that document
to the terms of the GNU GPL.

Today the Wen Quan Yi font collection includes hand-drawn CJK ideographs
in 9pt, 10pt, 11pt, and 12pt as well as a vector font.  The 12pt type
has glyphs the same size as the GNU Unifont, except it adds two more
pixels of leading (blank space) below each line.  This makes CJK easier
to read.  For these reasons, Wen Quan Yi is recommended over the GNU Unifont
for anyone planning to primarily use CJK ideographs.

See http://wqy.sourceforge.net/cgi-bin/enindex.cgi (English) or
http://wenq.org (Chinese) for more information on Wen Quan Yi.

In 2006, the GNU Unifont package was orphaned in Debian.

In late 2007, Paul Hardy became interested in adding to the GNU Unifont.
He wrote a couple of programs to convert GNU Unifont .hex files to and
from bitmap images for easy editing with any graphics software.  He began
by combining the latest glyphs available for the GNU Unifont.  This
starting point was posted at http://czyborra.com as the 2007-12-31 version
of unifont.hex.  Shortly after that, Roman Czyborra's website went down.
Paul Hardy then started posting complete copies of GNU Unifont on his
website, "http://unifoundry.com/unifont.html".

Roman Czyborra encouraged Paul Hardy to continue this work on GNU Unifont.

In early 2008, Paul Hardy learned of Qianqian Fang's work.  Qianqian
encouraged a combining of effort, and Paul Hardy at that point created
two versions of the GNU Unifont: one with the original Chinese ideographs
(which Roman Czyborra copied from a Japanese font in the public domain),
and one with Qianqian Fang's Wen Quan Yi (Spring of Letters) ideographs.
The Wen Quan Yi font provides far more coverage of CJK ideographs than
the original Japanese font did, and is of higher quality.

Paul Hardy created a version of both the font with the original CJK
ideographs from Japan and with CJK ideographs from Wen Quan Yi that
contained combining circles.  He then wrote a post-processing program
to remove the combining circles from the final font.

In 2005, Luis Gonzalez Miranda (http://www.lgm.cl) created a set of
Fontforge scripts and Perl programs.  Paul Hardy modified Luis' software
in 2008 to cover the full Unicode 5.1 Basic Multilingual Plane range.
Luis gave Paul Hardy permission to release this modified version under
the terms of "the GNU General Public License, version 2 or (at your
option) a later version."

On 4 July 2008, Paul Hardy was looking through all of Roman Czyborra's
Perl scripts.  One of these, "braille", contained a comment from 2003
that the original GNU Unifont did not generate its Braille patterns
(U+2800..U+28FF) correctly.  The modified script fixed that bug.  Paul
Hardy incorporated the correct Braile glyphs into the 6 July 2008
release of the GNU Unifont.  All previous versions probably contain
this Braille bug and should be replaced.

Other notable additions include:

     - Incorporation of CJK glyphs from Qianqian Fang's fonts
     - Incorporation of Rich Felker's Tibetan glyphs
     - Replacement of the Hangul Syllables block with a thin stroke font
       (Roman had mentioned wanting to do this someday on his website)
     - Addition of circled pencil glyphs for the Private Use Area
       (suggested as an accpetable rendering in the Unicode 5.0 Standard)
     - Proper handling of combining characters in the TrueType version
     - Proper handling of space glyphs in the TrueType version

The hex2bdf script in this release is Roman's original script, not the
modified version that produced two BDF files (one for 8 pixel wide glyphs
and another for 16 pixel wide glyphs).  The TrueType font should be used

debian/changelog
unifont (1:5.1.20080914-1.3) unstable; urgency=low

  * Non-maintainer upload.
  * debian/rules: Split and inline "install" target into "binary-arch" and
    "binary-indep" and adjust dependencies.  (Without this,
    "dpkg-buildpackage -B" ends up running the build-indep rules and dying
    on the buildds.)
  * debian/rules: Also fix so dpkg-buildpackage -j works, and only builds
    the font files once.

 -- Samuel Bronson <naesten@gmail.com>  Sat, 12 Jan 2013 13:09:44 -0500

unifont (1:5.1.20080914-1.2) unstable; urgency=low

  [ Samuel Bronson ]
  * Non-maintainer upload.
  * Actually build the font files, using build-arch and build-indep to
    avoid killing any buildds. (Closes: #697902)
  * Don't include bogus glyphs for unallocated (as of Unicode 5.1) or
    private-use characters. (Closes: #584672)

  [ Paul Wise ]
  * Place the newly compiled fonts in precompiled dir after build.

 -- Samuel Bronson <naesten@gmail.com>  Thu, 10 Jan 2013 20:15:33 -0500

unifont (1:5.1.20080914-1.1) unstable; urgency=low

  * Non-maintainer upload to achieve release goal of getting rid
    of defoma.

 -- Christian Perrier <bubulle@debian.org>  Wed, 20 Jul 2011 22:14:06 +0200

unifont (1:5.1.20080914-1) unstable; urgency=low

  * Changed unifont.ttf location in defoma-hints file. (Closes: #498874)
  * Modified src/hex2sfd to superimpose combining diacritical marks on
    top of preceding glyphs in TrueType font as per the Unicode Standard.
    (Closes: #497326)
  * Modified "cp -a" ==> "cp -p" and eliminated "install -t" in Makefiles
    for BSD Unix compatibility.
  * Renamed "combining.txt" to "combining.dat" because it is a data file.
  * Generated new font/precompiled/unifont.ttf font from the above changes.

 -- Paul Hardy <unifoundry@unifoundry.com>  Sun,  7 Sep 2008 23:09:17 -0700

unifont (1:5.1.20080820-1) unstable; urgency=low

  * Replaced Hangul Syllables block with glyphs in unifont.bdf file
    in Debian bf-utf-source package, originally from the Debian
    xfonts-baekmuk package (Closes: #495729)
    - Avoids license ambiguity of unifont's original Hanterm fonts
    - Brings two diverging Debian copies of the unifont font closer together
  * Deleted references to and all derivative copies of any Hanterm fonts
  * Changed unifont package description in debian/control to mention
    that the package now installs /usr/share/unifont/unifont.hex and
    other files and is no longer just a metapackage (Closes: #494662)

 -- Paul Hardy <unifoundry@unifoundry.com>  Wed, 20 Aug 2008 00:08:15 -0700

unifont (1:5.1.20080808-2) unstable; urgency=low

  * Install unifont.hex [and other files while I was at it] in
    /usr/share/unifont for GRUB (Closes: #494473)
    - unifont.hex provides glyphs with no combining circles, etc. and
      is the same file as font/compiled/unifont-5.1.20080808.hex
    - unifontfull.hex provides glyphs with combining circles, etc. and
      is the same file as font/compiled/unifontfull-5.1.20080808.hex.
      If editing hex glyphs, this file is the master version to edit;
      unifont.hex is generated from this full version -- see uniunmask(1)
      for details.
    - combining.txt lists all Unicode 5.1 code points that are combining
      glyphs, and hence should have zero-width.
    - README gives a brief description of the above three files
  * Corrected typos in the previous changelog entry.

 -- Paul Hardy <unifoundry@unifoundry.com>  Sat,  9 Aug 2008 20:55:27 -0700

unifont (1:5.1.20080808-1) unstable; urgency=low

  * Added needed Replaces, Conflicts lines in debian/control (Closes: #493711)
  * Added "Build-Depends-Indep: defoma" line in debian/control for access
    to dh_installdefoma to install TrueType font (Closes: #494245)
  * Changed ttf-unifont.defoma-hints file entry for "GeneralFamily"
    from "SansSerif" to "Typewriter" as a better match
  * Changed dh_lintian lines in debian/rules to use "( ... ) || /bin/true"
    so debuild would succeed on a generic Etch DVD installation that
    doesn't include dh_lintian in the debhelper package
  * Added updates from Ubuntu unifont package:
    - hex2bdf now uses additional BDF header lines for better font
      description
    - hex2bdf-split (originally Dave Starner's adaptation of Roman
      Czyborra's hex2bdf) renamed from hex2bdf.  Use is discouraged
      but it's still here for now for completeness; wrote man page
      & changed name of output font from "-gnu-fixed-*" to "-gnu-unifont-*"
    - Added Georgian glyphs contributed to Ubuntu's unifont package by
      Gia Shervashidze and Vladimer Sichinava
  * Merged man page & Makefile changes done by Anthony Fok from Debian
    diff file into .orig.tar.gz
  * Made minor changes to existing programs in ./src
    - unibmp2hex.c: forced double-wide glyphs where appropriate in
      U+FFxx range
    - uniunmask.c: increased buffer size & hard-coded "wide" variable
      to avoid accidentally invoking a half-implemented extended
      character widths
    - unihex2bmp.c: same changes as in uniunmask.c
  * Added to existing documentation
    - Updated "./README" to reflect changes to Makefiles and add
      more information
    - Added "./font/hexsrc/README" to describe how to use all files
      in that directory to create a font
    - Added full font embedding exception to GPL from Wen Quan Yi
  * Added improved glyphs for the U+FF00..U+FFFF block

src/unihex2bmp.c
/*
   unihex2bmp - program to turn a GNU Unifont hex glyph page of 256 code
                points into a Microsoft Bitmap (.bmp) or Wireless Bitmap file

   Synopsis: unihex2bmp [-iin_file.hex] [-oout_file.bmp]
                [-f] [-phex_page_num] [-w]


   Author: Paul Hardy, unifoundry <at> unifoundry.com, December 2007
   
   
    Copyright (C) 2007-2008 Paul Hardy

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAXBUF 256


/*
   These are the GNU Unifont hex strings for '0'-'9' and 'A'-'F',
   for encoding as bit strings in row and column headers.

   Looking at the final bitmap as a grid of 32*32 bit tiles, the
   first row contains a hexadecimal character string of the first
   3 hex digits in a 4 digit Unicode character name; the top column
   contains a hex character string of the 4th (low-order) hex digit
   of the Unicode character.
*/
char *hex[18]= {
      "0030:00000000182442424242424224180000",  /* Hex digit 0 */
      "0031:000000000818280808080808083E0000",  /* Hex digit 1 */
      "0032:000000003C4242020C102040407E0000",  /* Hex digit 2 */
      "0033:000000003C4242021C020242423C0000",  /* Hex digit 3 */
      "0034:00000000040C142444447E0404040000",  /* Hex digit 4 */
      "0035:000000007E4040407C020202423C0000",  /* Hex digit 5 */
      "0036:000000001C2040407C424242423C0000",  /* Hex digit 6 */
      "0037:000000007E0202040404080808080000",  /* Hex digit 7 */
      "0038:000000003C4242423C424242423C0000",  /* Hex digit 8 */
      "0039:000000003C4242423E02020204380000",  /* Hex digit 9 */
      "0041:0000000018242442427E424242420000",  /* Hex digit A */
      "0042:000000007C4242427C424242427C0000",  /* Hex digit B */
      "0043:000000003C42424040404042423C0000",  /* Hex digit C */
      "0044:00000000784442424242424244780000",  /* Hex digit D */
      "0045:000000007E4040407C404040407E0000",  /* Hex digit E */
      "0046:000000007E4040407C40404040400000",  /* Hex digit F */
      "0055:000000004242424242424242423C0000",  /* Unicode 'U' */
      "002B:0000000000000808087F080808000000"   /* Unicode '+' */
   };
unsigned char hexbits[18][32]; /* The above digits converted into bitmap */

unsigned unipage=0;        /* Unicode page number, 0x00..0xff */
int flip=0;                /* transpose entire matrix as in Unicode book */



int main(int argc, char *argv[]) {

   int i, j;                  /* loop variables                    */
   unsigned k0;               /* temp Unicode char variable        */
   unsigned swap;             /* temp variable for swapping values */
   char inbuf[256];           /* input buffer                      */
// unsigned headersize=0x28;  /* size of header (fixed in Windows) */
   unsigned filesize;         /* size of file in bytes             */
   unsigned bitmapsize;       /* size of bitmap image in bytes     */
   unsigned thischar;         /* the current character             */
   unsigned char thischarbyte; /* unsigned char lowest byte of Unicode char */
   int thischarrow;           /* row 0..15 where this character belongs  */
   int thiscol;               /* column 0..15 where this character belongs */
   int toppixelrow;           /* pixel row, 0..16*32-1               */
   unsigned lastpage=0;       /* the last Unicode page read in font file */
   int wbmp=0;                /* set to 1 if writing .wbmp format file */

   unsigned char bitmap[17*32][18*4]; /* final bitmap */
   unsigned char charbits[32][4];  /* bitmap for one character, 4 bytes/row */

   char *infile="", *outfile="";  /* names of input and output files */
   FILE *infp, *outfp;      /* file pointers of input and output files */

   int init();                  /* initializes bitmap row/col labeling, &c. */
   int hex2bit();               /* convert hex string --> bitmap */

   bitmapsize = 17*32*18*4;  /* 17 rows by 18 cols, each 4 bytes */

   if (argc > 1) {
      for (i = 1; i < argc; i++) {
         if (argv[i][0] == '-') {  /* this is an option argument */
            switch (argv[i][1]) {
               case 'f':  /* flip (transpose) glyphs in bitmap as in standard */
                  flip = 1;
                  break;
               case 'i':  /* name of input file */
                  infile = &argv[i][2];
                  break;
               case 'o':  /* name of output file */
                  outfile = &argv[i][2];
                  break;
               case 'p':  /* specify a Unicode page other than default of 0 */
                  sscanf(&argv[i][2], "%x", &unipage); /* Get Unicode page */
                  break;
               case 'w':  /* write a .wbmp file instead of a .bmp file */
                  wbmp = 1;
                  break;
               default:   /* if unrecognized option, print list and exit */
                  fprintf(stderr, "\nSyntax:\n\n");
                  fprintf(stderr, "   %s -p<Unicode_Page> ", argv[0]);
                  fprintf(stderr, "-i<Input_File> -o<Output_File> -w\n\n");
                  fprintf(stderr, "   -w specifies .wbmp output instead of ");
                  fprintf(stderr, "default Windows .bmp output.\n\n");
                  fprintf(stderr, "   -p is followed by 1 to 6 ");
                  fprintf(stderr, "Unicode page hex digits ");
                  fprintf(stderr, "(default is Page 0).\n\n");
                  fprintf(stderr, "\nExample:\n\n");
                  fprintf(stderr, "   %s -p83 -iunifont.hex -ou83.bmp\n\n\n",
                         argv[0]);
                  exit(1);
            }
         }
      }
   }
   /*
      Make sure we can open any I/O files that were specified before
      doing anything else.
   */
   if (strlen(infile) > 0) {
      if ((infp = fopen(infile, "r")) == NULL) {
         fprintf(stderr, "Error: can't open %s for input.\n", infile);
         exit(1);
      }
   }
   else {
      infp = stdin;
   }
   if (strlen(outfile) > 0) {
      if ((outfp = fopen(outfile, "w")) == NULL) {
         fprintf(stderr, "Error: can't open %s for output.\n", outfile);
         exit(1);
      }
   }
   else {
      outfp = stdout;
   }

   (void)init(bitmap); /* initialize bitmap with row/column headers, etc. */

   /*
      Read in the characters in the page
   */
   while (lastpage <= unipage && fgets(inbuf, MAXBUF-1, infp) != NULL) {
      sscanf(inbuf, "%x", &thischar);
      lastpage = thischar >> 8; /* keep Unicode page to see if we can stop */
      if (lastpage == unipage) {
         thischarbyte = (unsigned char)(thischar & 0xff);
         for (k0=0; inbuf[k0] != ':'; k0++);
         k0++;
         hex2bit(&inbuf[k0], charbits);  /* convert hex string to 32*4 bitmap */

         /*
            Now write character bitmap upside-down in page array, to match
            .bmp file order.  In the .wbmp` and .bmp files, white is a '1'
            bit and black is a '0' bit, so complement charbits[][].
         */

         thiscol = (thischarbyte & 0xf) + 2;  /* column number will be 1..16  */
         thischarrow = thischarbyte >> 4;     /* charcter row number, 0..15   */
         if (flip) {  /* swap row and column placement */
            swap = thiscol;
            thiscol = thischarrow;
            thischarrow = swap;
            thiscol += 2;       /* column index starts at 1 */
            thischarrow -= 2;   /* row index starts at 0    */
         }
         toppixelrow = 32 * (thischarrow + 1) - 1; /* from bottom to top    */
   
         /*
            Copy the center of charbits[][] because hex characters only
            occupy rows 8 to 23 and column byte 2 (and for 16 bit wide
            characters, byte 3).  The charbits[][] array was given 32 rows
            and 4 column bytes for completeness in the beginning.
         */
         for (i=8; i<24; i++) {
            bitmap[toppixelrow + i][(thiscol << 2) | 1] =
               ~charbits[i][1] & 0xff;
            bitmap[toppixelrow + i][(thiscol << 2) | 2] =
               ~charbits[i][2] & 0xff;
         }
      }
   }
   /*
      Now write the appropriate bitmap file format, either
      Wireless Bitmap or Microsoft Windows bitmap.
   */
   if (wbmp) {  /* Write a Wireless Bitmap .wbmp format file */
      /*
         Write WBMP header
      */
      fprintf(outfp, "%c", 0x00); /* Type of image; always 0 (monochrome) */
      fprintf(outfp, "%c", 0x00); /* Reserved; always 0                   */
      fprintf(outfp, "%c%c", 0x84, 0x40); /* Width  = 576 pixels          */
      fprintf(outfp, "%c%c", 0x84, 0x20); /* Height = 544 pixels          */
      /*
         Write bitmap image
      */
      for (toppixelrow=0; toppixelrow <= 17*32-1; toppixelrow++) {
         for (j=0; j<18; j++) {
            fprintf(outfp, "%c", bitmap[toppixelrow][(j<<2)    ]);
            fprintf(outfp, "%c", bitmap[toppixelrow][(j<<2) | 1]);
            fprintf(outfp, "%c", bitmap[toppixelrow][(j<<2) | 2]);
            fprintf(outfp, "%c", bitmap[toppixelrow][(j<<2) | 3]);
         }
      }
   }
   else {  /* otherwise, write a Microsoft Windows .bmp format file */
      /*
         Write the .bmp file -- start with the header, then write the bitmap
      */

      /* 'B', 'M' appears at start of every .bmp file */
      fprintf(outfp, "%c%c", 0x42, 0x4d);

      /* Write file size in bytes */
      filesize   = 0x3E + bitmapsize;
      fprintf(outfp, "%c", (unsigned char)((filesize        ) & 0xff));
      fprintf(outfp, "%c", (unsigned char)((filesize >> 0x08) & 0xff));
      fprintf(outfp, "%c", (unsigned char)((filesize >> 0x10) & 0xff));
      fprintf(outfp, "%c", (unsigned char)((filesize >> 0x18) & 0xff));

      /* Reserved - 0's */
      fprintf(outfp, "%c%c%c%c", 0x00, 0x00, 0x00, 0x00);

      /* Offset from start of file to bitmap data */
      fprintf(outfp, "%c%c%c%c", 0x3E, 0x00, 0x00, 0x00);

      /* Length of bitmap info header */
      fprintf(outfp, "%c%c%c%c", 0x28, 0x00, 0x00, 0x00);

      /* Width of bitmap in pixels */
      fprintf(outfp, "%c%c%c%c", 0x40, 0x02, 0x00, 0x00);

      /* Height of bitmap in pixels */
      fprintf(outfp, "%c%c%c%c", 0x20, 0x02, 0x00, 0x00);

      /* Planes in bitmap (fixed at 1) */
      fprintf(outfp, "%c%c", 0x01, 0x00);

      /* bits per pixel (1 = monochrome) */
      fprintf(outfp, "%c%c", 0x01, 0x00);

      /* Compression (0 = none) */
      fprintf(outfp, "%c%c%c%c", 0x00, 0x00, 0x00, 0x00);

      /* Size of bitmap data in bytes */
      fprintf(outfp, "%c", (unsigned char)((bitmapsize        ) & 0xff));
      fprintf(outfp, "%c", (unsigned char)((bitmapsize >> 0x08) & 0xff));
      fprintf(outfp, "%c", (unsigned char)((bitmapsize >> 0x10) & 0xff));
      fprintf(outfp, "%c", (unsigned char)((bitmapsize >> 0x18) & 0xff));

      /* Horizontal resolution in pixels per meter */
      fprintf(outfp, "%c%c%c%c", 0xC4, 0x0E, 0x00, 0x00);

      /* Vertical resolution in pixels per meter */
      fprintf(outfp, "%c%c%c%c", 0xC4, 0x0E, 0x00, 0x00);

      /* Number of colors used */
      fprintf(outfp, "%c%c%c%c", 0x02, 0x00, 0x00, 0x00);

      /* Number of important colors */
      fprintf(outfp, "%c%c%c%c", 0x02, 0x00, 0x00, 0x00);

      /* The color black: B=0x00, G=0x00, R=0x00, Filler=0xFF */
      fprintf(outfp, "%c%c%c%c", 0x00, 0x00, 0x00, 0x00);

      /* The color white: B=0xFF, G=0xFF, R=0xFF, Filler=0xFF */
      fprintf(outfp, "%c%c%c%c", 0xFF, 0xFF, 0xFF, 0x00);

      /*
         Now write the raw data bits.  Data is written from the lower
         left-hand corner of the image to the upper right-hand corner
         of the image.
      */
      for (toppixelrow=17*32-1; toppixelrow >= 0; toppixelrow--) {
         for (j=0; j<18; j++) {
            fprintf(outfp, "%c", bitmap[toppixelrow][(j<<2)    ]);
            fprintf(outfp, "%c", bitmap[toppixelrow][(j<<2) | 1]);
            fprintf(outfp, "%c", bitmap[toppixelrow][(j<<2) | 2]);

            fprintf(outfp, "%c", bitmap[toppixelrow][(j<<2) | 3]);
         }
      }
   }
   exit(0);
}

/*
   Convert the portion of a hex string after the ':' into a character bitmap.

   If string is >= 128 characters, it will fill all 4 bytes per row.
   If string is >= 64 characters and < 128, it will fill 2 bytes per row.
   Otherwise, it will fill 1 byte per row.
*/
int hex2bit(char *instring, unsigned char character[32][4]) {

   int i;  /* current row in bitmap character */
   int j;  /* current character in input string */
   int k;  /* current byte in bitmap character */
   int width;  /* number of output bytes to fill: 1, 2, or 4 */

   for (i=0; i<32; i++)  /* erase previous character */
      character[i][0] = character[i][1] = character[i][2] = character[i][3] = 0; 
   j=0;  /* current location is at beginning of instring */

   // width = (strlen(instring) - 1) >> 4; /* 16 hex digits per 8 bytes */

   if (strlen(instring) <= 34)  /* 32 + possible '\r', '\n' */
      width = 0;
   else if (strlen(instring) <= 66)  /* 64 + possible '\r', '\n' */
      width = 1;
   else
      width = 4;

   k = (width > 1) ? 0 : 1;  /* if width < 3, start at index 1 else at 0 */

   for (i=8; i<24; i++) {  /* 16 rows per input character, rows 8..23 */
      sscanf(&instring[j], "%2hhx", &character[i][k]);
      j += 2;
      if (width > 0) { /* add next pair of hex digits to this row */
         sscanf(&instring[j], "%2hhx", &character[i][k+1]);
         j += 2;
      }
      if (width > 1) { /* add 2 next pairs of hex digits to this row */
         sscanf(&instring[j], "%2hhx", &character[i][k+2]);
         j += 2;
         sscanf(&instring[j], "%2hhx", &character[i][k+3]);
         j += 2;
      }
   }

   return(0);
}

int init(unsigned char bitmap[17*32][18*4]) {
   int i, j;
   unsigned char charbits[32][4];  /* bitmap for one character, 4 bytes/row */
   unsigned toppixelrow;
   unsigned thiscol;
   unsigned char pnybble0, pnybble1, pnybble2, pnybble3;

   for (i=0; i<18; i++) { /* bitmaps for '0'..'9', 'A'-'F', 'u', '+' */

      hex2bit(&hex[i][5], charbits);  /* convert hex string to 32*4 bitmap */

      for (j=0; j<32; j++) hexbits[i][j] = ~charbits[j][1];
   }

   /*
      Initialize bitmap to all white.
   */
   for (toppixelrow=0; toppixelrow < 17*32; toppixelrow++) {
      for (thiscol=0; thiscol<18; thiscol++) {
         bitmap[toppixelrow][(thiscol << 2)    ] = 0xff;
         bitmap[toppixelrow][(thiscol << 2) | 1] = 0xff;
         bitmap[toppixelrow][(thiscol << 2) | 2] = 0xff;
         bitmap[toppixelrow][(thiscol << 2) | 3] = 0xff;
      }
   }
   /*
      Write the "u+nnnn" table header in the upper left-hand corner,
      where nnnn is the upper 16 bits of a 32-bit Unicode assignment.
   */
   pnybble3 = (unipage >> 20);
   pnybble2 = (unipage >> 16) & 0xf;
   pnybble1 = (unipage >> 12) & 0xf;
   pnybble0 = (unipage >>  8) & 0xf;
   for (i=0; i<32; i++) {
      bitmap[i][1] = hexbits[16][i];  /* copy 'u' */
      bitmap[i][2] = hexbits[17][i];  /* copy '+' */
      bitmap[i][3] = hexbits[pnybble3][i];
      bitmap[i][4] = hexbits[pnybble2][i];
      bitmap[i][5] = hexbits[pnybble1][i];
      bitmap[i][6] = hexbits[pnybble0][i];
   }
   /*
      Write low-order 2 bytes of Unicode number assignments, as hex labels
   */
   pnybble3 = (unipage >> 4) & 0xf;  /* Highest-order hex digit */
   pnybble2 = (unipage     ) & 0xf;  /* Next highest-order hex digit */
   /*
      Write the column headers in bitmap[][] (row headers if flipped)
   */
   toppixelrow = 32 * 17 - 1; /* maximum pixel row number */
   /*
      Label the column headers. The hexbits[][] bytes are split across two
      bitmap[][] entries to center a the hex digits in a column of 4 bytes.
      OR highest byte with 0xf0 and lowest byte with 0x0f to make outer
      nybbles white (0=black, 1-white).
   */
   for (i=0; i<16; i++) {
      for (j=0; j<32; j++) {
         if (flip) {  /* transpose matrix */
            bitmap[j][((i+2) << 2) | 0]  = (hexbits[pnybble3][j] >> 4) | 0xf0;
            bitmap[j][((i+2) << 2) | 1]  = (hexbits[pnybble3][j] << 4) |
                                           (hexbits[pnybble2][j] >> 4);
            bitmap[j][((i+2) << 2) | 2]  = (hexbits[pnybble2][j] << 4) |
                                           (hexbits[i][j] >> 4);
            bitmap[j][((i+2) << 2) | 3]  = (hexbits[i][j] << 4) | 0x0f;
         }
         else {
            bitmap[j][((i+2) << 2) | 1] = (hexbits[i][j] >> 4) | 0xf0;
            bitmap[j][((i+2) << 2) | 2] = (hexbits[i][j] << 4) | 0x0f;
         }
      }
   }
   /*
      Now use the single hex digit column graphics to label the row headers.
   */
   for (i=0; i<16; i++) {
      toppixelrow = 32 * (i + 1) - 1; /* from bottom to top    */

      for (j=0; j<32; j++) {
         if (!flip) {  /* if not transposing matrix */
            bitmap[toppixelrow + j][4] = hexbits[pnybble3][j];
            bitmap[toppixelrow + j][5] = hexbits[pnybble2][j];
         }
         bitmap[toppixelrow + j][6] = hexbits[i][j];
      }
   }
   /*
      Now draw grid lines in bitmap, around characters we just copied.
   */
   /* draw vertical lines 2 pixels wide */
   for (i=1*32; i<17*32; i++) {
      if ((i & 0x1f) == 7)
         i++;
      else if ((i & 0x1f) == 14)
         i += 2;
      else if ((i & 0x1f) == 22)
         i++;
      for (j=1; j<18; j++) {
         bitmap[i][(j << 2) | 3] &= 0xfe;
      }
   }
   /* draw horizontal lines 2 pixels tall */
   for (i=1*32-1; i<18*32-1; i+=32) {
      for (j=2; j<18; j++) {
         bitmap[i][(j << 2)    ] = 0x00;
         bitmap[i][(j << 2) | 1] = 0x81;
         bitmap[i][(j << 2) | 2] = 0x81;
         bitmap[i][(j << 2) | 3] = 0x00;
      }
   }
   /* fill in top left corner pixel of grid */
   bitmap[31][7] = 0xfe;

   return(0);
}
